internal/messaging/response_handler.go

```diff
@@ -1,117 +1,539 @@
+		addErrorToResponseHeader(response, fmt.Sprintf("error loading ABI: %v", err))
 		return true
 	}
-	price, err := strconv.Atoi(response.MintResponse.Price.Value)
+	address := crypto.PubkeyToAddress(h.pk.ToECDSA().PublicKey)
+
+	bookingTokenAddress := common.HexToAddress(h.cfg.BookingTokenAddress)
+
+	packedData, err := abi.Pack("getSupplierName", address)
+	if err != nil {
+		h.logger.Infof("Error packing data for getSupplierName: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error packing data: %v", err))
+		return true
+	}
+
+	msg := ethereum.CallMsg{
+		To:   &bookingTokenAddress,
+		Data: packedData,
+	}
+	result, err := h.ethClient.CallContract(context.Background(), msg, nil)
 	if err != nil {
-		addErrorToResponseHeader(response, fmt.Sprintf("error parsing price value: %v", err))
+		h.logger.Infof("Error calling contract: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error calling contract: %v", err))
 		return true
 	}
-	success, txID, err := h.tvmClient.SendTxAndWait(ctx, createNFTAction(owner, buyer, uint64(response.MintResponse.BuyableUntil.Seconds), uint64(price), response.MintResponse.MintId))
+
+	// Unpack the result
+	var supplierName string
+	err = abi.UnpackIntoInterface(&supplierName, "getSupplierName", result)
+	if err != nil {
+		h.logger.Infof("Error unpacking result: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error unpacking result: %v", err))
+		return true
+	}
+
+	if supplierName != h.cfg.SupplierName {
+		h.logger.Debugf("Not registered with correct supplier name: %v != %v", supplierName, h.cfg.SupplierName)
+		h.logger.Debugf("Registerring with supplierName: %v", h.cfg.SupplierName)
+		err := register(h.ethClient, abi, h.pk.ToECDSA(), h.cfg.SupplierName, bookingTokenAddress)
+		if err != nil {
+			addErrorToResponseHeader(response, fmt.Sprintf("error registering supplier: %v", err))
+			return true
+		}
+	} else {
+		h.logger.Debugf("Supplier is already registered with supplierName: %v", supplierName)
+	}
+
+	// owner := h.ethClient.Address()
+	// if response.MintResponse.Header == nil {
+	// 	response.MintResponse.Header = &typesv1alpha.ResponseHeader{}
+	// }
+	// TODO @evlekht ensure that request.MintRequest.BuyerAddress is c-chain address format, not x/p/t chain
+	buyer := common.HexToAddress(request.MintRequest.BuyerAddress)
+	// price, err := strconv.Atoi(response.MintResponse.Price.Value)
+	// if err != nil {
+	// 	addErrorToResponseHeader(response, fmt.Sprintf("error parsing price value: %v", err))
+	// 	return true
+	// }
+
+	h.logger.Debugf("abi: %v", abi)
+
+	// TODO @evlekht unhardocoded, figure out what it is at all
+	// uri := "data:application/json;base64,eyJuYW1lIjoiQ2FtaW5vIE1lc3NlbmdlciBCb29raW5nVG9rZW4gVGVzdCJ9Cg=="
+
+	// Get a Token URI for the token.
+	jsonPlain, uri, err := createTokenURIforMintResponse(response.MintResponse)
+	if err != nil {
+		addErrorToResponseHeader(response, fmt.Sprintf("error creating token URI: %v", err))
+		return true
+	}
+
```

Comment:
I think it's better to tidy up the code before merging to main branch. We should remove redundant comments/code/todos.